File: frontend/mobile/lib/models/relationship.dart
class Relationship {
  final String id;
  final String userId;
  final String counterpartyName;
  final String? counterpartyId;
  final String? relationshipType;
  final String? notes;
  final double totalGiven;
  final double totalReceived;
  final double balance;
  final DateTime? lastTransactionDate;
  final DateTime createdAt;

  Relationship({
    required this.id,
    required this.userId,
    required this.counterpartyName,
    this.counterpartyId,
    this.relationshipType,
    this.notes,
    required this.totalGiven,
    required this.totalReceived,
    required this.balance,
    this.lastTransactionDate,
    required this.createdAt,
  });

  factory Relationship.fromJson(Map<String, dynamic> json) => Relationship(
        id: json['id'],
        userId: json['user_id'],
        counterpartyName: json['counterparty_name'],
        counterpartyId: json['counterparty_id'],
        relationshipType: json['relationship_type'],
        notes: json['notes'],
        totalGiven: (json['total_given'] as num).toDouble(),
        totalReceived: (json['total_received'] as num).toDouble(),
        balance: (json['balance'] as num).toDouble(),
        lastTransactionDate: json['last_transaction_date'] != null
            ? DateTime.tryParse(json['last_transaction_date'])
            : null,
        createdAt: DateTime.parse(json['created_at']),
      );

  /// 양수: 내가 더 줌 / 음수: 내가 더 받음
  bool get iGaveMore => balance > 0;
}

File: backend/app/schemas/relationship.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field, computed_field


class RelationshipCreate(BaseModel):
    counterparty_name: str = Field(..., min_length=1, max_length=100)
    relationship_type: Optional[str] = Field(None, max_length=50)
    notes: Optional[str] = None


class RelationshipUpdate(BaseModel):
    relationship_type: Optional[str] = Field(None, max_length=50)
    notes: Optional[str] = None


class RelationshipResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: uuid.UUID
    counterparty_name: str
    counterparty_id: Optional[uuid.UUID]
    relationship_type: Optional[str]
    notes: Optional[str]
    total_given: Decimal
    total_received: Decimal
    last_transaction_date: Optional[datetime]
    created_at: datetime

    @computed_field
    @property
    def balance(self) -> Decimal:
        """양수: 내가 더 줌 / 음수: 내가 더 받음"""
        return self.total_given - self.total_received
